service/storage/operations_test.go

```diff
@@ -0,0 +1,321 @@
+		assert.Equal(t, append(testEvents1, testEvents2...), got)
+	})
+
+	t.Run("retrieve events does not include types not asked for", func(t *testing.T) {
+		txn := db.NewTransaction(false)
+
+		var got []flow.Event
+		err := RetrieveEvents(42, []flow.EventType{testTyp1, "another-type"}, &got)(txn)
+
+		assert.NoError(t, err)
+		assert.Equal(t, testEvents1, got)
+	})
+
+	t.Run("retrieve events does not include types not asked for", func(t *testing.T) {
+		txn := db.NewTransaction(false)
+
+		var got []flow.Event
+		err := RetrieveEvents(42, []flow.EventType{testTyp2, "another-type"}, &got)(txn)
+
+		assert.NoError(t, err)
+		assert.Equal(t, testEvents2, got)
+	})
+}
+
+func TestSaveAndRetrieve_Payload(t *testing.T) {
+	db := helpers.InMemoryDB(t)
+	defer db.Close()
+
+	path, _ := ledger.ToPath([]byte("aac513eb1a0457700ac3fa8d292513e1"))
+	key := ledger.NewKey([]ledger.KeyPart{
+		ledger.NewKeyPart(0, []byte(`owner`)),
+		ledger.NewKeyPart(1, []byte(`controller`)),
+		ledger.NewKeyPart(2, []byte(`key`)),
+	})
+	payload1 := ledger.NewPayload(
+		key,
+		ledger.Value(`test1`),
+	)
+	payload2 := ledger.NewPayload(
+		key,
+		ledger.Value(`test2`),
+	)
+
+	t.Run("save two different payloads for same path at different heights", func(t *testing.T) {
+		txn := db.NewTransaction(true)
+
+		err := SavePayload(42, path, payload1)(txn)
+
+		assert.NoError(t, err)
+
+		err = SavePayload(84, path, payload2)(txn)
+
+		assert.NoError(t, err)
+
+		err = txn.Commit()
+		require.NoError(t, err)
+	})
+
+	t.Run("retrieve payload at its first indexed height", func(t *testing.T) {
+		txn := db.NewTransaction(false)
+
+		var got ledger.Payload
+		err := RetrievePayload(42, path, &got)(txn)
+
+		assert.NoError(t, err)
+		assert.Equal(t, *payload1, got)
+	})
+
+	t.Run("retrieve payload at its second indexed height", func(t *testing.T) {
+		txn := db.NewTransaction(false)
+
+		var got ledger.Payload
+		err := RetrievePayload(84, path, &got)(txn)
+
+		assert.NoError(t, err)
+		assert.Equal(t, *payload2, got)
+	})
+
+	t.Run("retrieve payload between first and second indexed height", func(t *testing.T) {
+		txn := db.NewTransaction(false)
+
+		var got ledger.Payload
+		err := RetrievePayload(63, path, &got)(txn)
+
+		assert.NoError(t, err)
+		assert.Equal(t, *payload1, got)
+	})
+
+	t.Run("retrieve payload before it was ever indexed", func(t *testing.T) {
+		txn := db.NewTransaction(false)
+
+		var got ledger.Payload
+		err := RetrievePayload(63, path, &got)(txn)
+
+		assert.NoError(t, err)
+		assert.Equal(t, *payload1, got)
+	})
+
+	t.Run("retrieve payload after last indexed", func(t *testing.T) {
+		txn := db.NewTransaction(false)
+
+		var got ledger.Payload
+		err := RetrievePayload(999, path, &got)(txn)
+
+		assert.NoError(t, err)
+		assert.Equal(t, *payload2, got)
+	})
+
+	t.Run("should fail if path does not match", func(t *testing.T) {
+		txn := db.NewTransaction(false)
+
+		var got ledger.Payload
+		err := RetrievePayload(42, ledger.Path{}, &got)(txn)
```

Comment:
This is dangerous, because `ledger.Path` is a byte slice in many past versions. You should explicitly initialize one to the usual length here, imo. There is a constant for `PathByteSize` in Flow Go.